File: tests/cpp/test_audio_groups.cpp
/**
 * Audio System Group Tests
 * 
 * Tests audio group operations including creation, volume control, and fading.
 */

#include "test_common.h"
#include "audio_manager.h"
#include <cmath>

using namespace audio;
using namespace std::chrono_literals;

void test_group_operations() {
    TEST("Audio Group Operations")
    
    auto& audio = AudioManager::GetInstance();
    
    // Create groups
    GroupHandle music = audio.CreateGroup();
    ASSERT(music.IsValid(), "Music group handle should be non-zero")
    
    GroupHandle sfx = audio.CreateGroup();
    ASSERT(sfx.IsValid(), "SFX group handle should be non-zero")
    ASSERT(sfx != music, "Group handles should be unique")
    
    // Test volume control
    audio.SetGroupVolume(music, 0.7f);
    ASSERT(std::abs(audio.GetGroupVolume(music) - 0.7f) < 0.01f, "Music group volume should be 0.7")
    
    audio.SetGroupVolume(sfx, 0.3f);
    ASSERT(std::abs(audio.GetGroupVolume(sfx) - 0.3f) < 0.01f, "SFX group volume should be 0.3")
    
    // Test fade (non-blocking, just verify it doesn't crash)
    audio.FadeGroup(music, 0.0f, 500ms);
    wait_ms(100);
    ASSERT(true, "Group fade operation should not crash")
    
    // Cleanup
    audio.DestroyGroup(music);
    audio.DestroyGroup(sfx);
    ASSERT(true, "Group destruction should not crash")
    
    END_TEST
}

int main(int argc, char* argv[]) {
    std::cout << "========================================" << std::endl;
    std::cout << "Audio System Group Tests" << std::endl;
    std::cout << "========================================" << std::endl;
    
    sound_dir = get_sound_dir(argc, argv);
    std::cout << "Sound directory: " << sound_dir << std::endl;
    
    // Initialize audio system
    auto& audio = AudioManager::GetInstance();
    if (!audio.Initialize()) {
        std::cerr << "FATAL: Failed to initialize audio system" << std::endl;
        return 1;
    }
    
    // Run all tests
    test_group_operations();
    
    // Final shutdown
    audio.Shutdown();
    
    // Print results
    std::cout << "\n========================================" << std::endl;
    std::cout << "Test Results" << std::endl;
    std::cout << "========================================" << std::endl;
    std::cout << "Passed: " << tests_passed << std::endl;
    std::cout << "Failed: " << tests_failed << std::endl;
    std::cout << "Total:  " << (tests_passed + tests_failed) << std::endl;
    
    if (tests_failed == 0) {
        std::cout << "\nALL TESTS PASSED ✓" << std::endl;
        return 0;
    } else {
        std::cout << "\nSOME TESTS FAILED ✗" << std::endl;
        return 1;
    }
}

File: tests/python/test_audio_groups.py
"""
Audio System Group Tests
Tests audio group operations including creation, volume control, and fading.
"""

from test_common import *

def test_group_operations():
    """Test: Audio group operations"""
    print("Test: Group operations... ", end="", flush=True)
    audio = game_audio.AudioManager.get_instance()
    audio.initialize()
    
    # Create groups
    music = audio.create_group("music")
    assert music.value != 0, "Music group handle should be non-zero"
    
    sfx = audio.create_group("sfx")
    assert sfx.value != 0, "SFX group handle should be non-zero"
    assert sfx.value != music.value, "Group handles should be unique"
    
    # Test volume control
    audio.set_group_volume(music, 0.7)
    assert abs(audio.get_group_volume(music) - 0.7) < 0.01, "Music group volume should be 0.7"
    
    audio.set_group_volume(sfx, 0.3)
    assert abs(audio.get_group_volume(sfx) - 0.3) < 0.01, "SFX group volume should be 0.3"
    
    # Test fade (non-blocking, just verify it doesn't crash)
    audio.fade_group(music, 0.0, timedelta(milliseconds=500))
    wait_ms(100)
    assert True, "Group fade operation should not crash"
    
    # Cleanup
    audio.destroy_group(music)
    audio.destroy_group(sfx)
    audio.shutdown()
    print("PASS")

def run_all_tests():
    """Run all group tests"""
    print("=== Python Audio Group Tests ===\n")
    
    tests = [
        test_group_operations,
    ]
    
    passed = 0
    failed = 0
    
    for test in tests:
        try:
            test()
            passed += 1
        except AssertionError as e:
            print(f"FAIL - {e}")
            failed += 1
        except Exception as e:
            print(f"ERROR - {e}")
            failed += 1
    
    print(f"\n=== Results: {passed} passed, {failed} failed ===")
    return failed == 0

if __name__ == "__main__":
    success = run_all_tests()
    sys.exit(0 if success else 1)
